nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core-api/src/main/java/org/apache/nifi/controller/AbstractConfiguredComponent.java

```diff
@@ -90,47 +107,80 @@ public void setAnnotationData(final String data) {
+                        final PropertyDescriptor descriptor = component.getPropertyDescriptor(entry.getKey());
+                        if (descriptor.isDynamicClasspathModifier() && !StringUtils.isEmpty(entry.getValue())) {
+                            final StandardPropertyValue propertyValue = new StandardPropertyValue(entry.getValue(), null, variableRegistry);
+                            modulePaths.add(propertyValue.evaluateAttributeExpressions().getValue());
                         }
                     }
+                }
 
-                    try {
-                        component.onPropertyModified(descriptor, oldValue, value);
-                    } catch (final Exception e) {
-                        // nothing really to do here...
-                    }
+                final boolean requiresInstanceClassLoading = ExtensionManager.requiresInstanceClassLoading(component.getClass().getTypeName());
+                if (requiresInstanceClassLoading) {
+                    processClasspathModifiers(modulePaths);
+                } else if (!requiresInstanceClassLoading && modulePaths.size() > 0) {
+                    // Component has property descriptors with dynamicallyModifiesClasspath set to true, but the class does not have @RequiresInstanceClassLoading
+                    logger.warn("One or more property descriptors intend to modify the classpath, " +
+                            "but component does not contain the {} annotation, classpath will not be modified",
+                            new Object[] {RequiresInstanceClassLoading.class.getName()});
                 }
             }
         } finally {
             lock.unlock();
         }
     }
 
+    // Keep setProperty/removeProperty private so that all calls go through setProperties
+    private void setProperty(final String name, final String value) {
+        if (null == name || null == value) {
+            throw new IllegalArgumentException();
+        }
+
+        final PropertyDescriptor descriptor = component.getPropertyDescriptor(name);
+
+        final String oldValue = properties.put(descriptor, value);
+        if (!value.equals(oldValue)) {
+
+            if (descriptor.getControllerServiceDefinition() != null) {
+                if (oldValue != null) {
+                    final ControllerServiceNode oldNode = serviceProvider.getControllerServiceNode(oldValue);
+                    if (oldNode != null) {
+                        oldNode.removeReference(this);
+                    }
+                }
+
+                final ControllerServiceNode newNode = serviceProvider.getControllerServiceNode(value);
+                if (newNode != null) {
+                    newNode.addReference(this);
+                }
+            }
+
+            try {
+                component.onPropertyModified(descriptor, oldValue, value);
+            } catch (final Exception e) {
+                // nothing really to do here...
```

Comment:
Hmm. . . So if user implements this method in Processor and it throws exception don't we have to at the very least notify the user that it failed?